### ISSUE TITLE
Cannot run on Pop!_OS: DllNotFoundException:

### ISSUE THREAD

**Issue by ghost** [aardvark-community/Aardvark.OpenImageDenoise#2]
I have just installed the NuGet package, and I'm trying to get this to work to see if I can have some denoising implemented into my RayTracer. At the moment all I want is to have something simple, but I can't even initialize the library it seems:
```
System.DllNotFoundException: Unable to load shared library 'OpenImageDenoise.dll' or one of its dependencies. In order to help diagnose loading problems, consider setting the LD_DEBUG environment variable: libOpenImageDenoise.dll: cannot open shared object file: No such file or directory
   at IntPtr Aardvark.OpenImageDenoise.OidnAPI.oidnNewDevice(DeviceFlags flags)
   at new Aardvark.OpenImageDenoise.Device(DeviceFlags flags, int numThreads) x 2
   at new RayTracer.Display.Dev.MainForm() in /home/rrm/Documents/Projects/RayTracer/RayTracer.Display.Dev/MainForm.cs:line 179
   at int RayTracer.Display.Dev.Program.Main(string[] args) in /home/rrm/Documents/Projects/RayTracer/RayTracer.Display.Dev/Program.cs:line 72
```

The line that's causing it:
```csharp

		//Denoise things
		Aardvark.Base.Aardvark.Init();
		denoiseDevice = new Device(); //<--------
```
I honestly  have no idea how to fix this since I can't find the `libOpenImageDenoise.dll` anywhere, and there are no instructions in the repo. I have tried fiddling around with the files in the `/Natives` folder in the repo, and by downloading natives from the OIDN official website, but it's still not working.

**Comment by luithefirst** [comment]
You were on the right track, you are missing the appropriate native assemblies for your OS. Unfortunately for this warpper we only have included winx64 so far as there was no need for another platform yet. `Aardvark.Init` would select the appropriate binaries and include them during runtime (most other libraries have linux and mac support). Still, if it cannot be found, there should be a fallback to look right next to your application executable. What is strange is that you get an error message where a dll is not found, because you are running Linux.

A quick solution you can try is to get the Linux binaries of the 1.3.0 release: https://github.com/OpenImageDenoise/oidn/releases/download/v1.3.0/oidn-1.3.0.x86_64.linux.tar.gz
and extract the \lib\ content next to your executable.

If you like to test thing from this repository, you can experiment with the test project.

As you can see here:
https://github.com/aardvark-community/Aardvark.OpenImageDenoise/blob/master/src/DenoiseTest/DenoiseTest.csproj
it directly includes the windows binaries (-> I removed in 94db950e). But actually Aardvark.Init also recognizes the startup directory to have a typical repository structure and registers the natives folder from there. I just ran a test and it seems to be working, but somehow saving the output images does not (-> was not saving to the expected directory, fixed in 94db950e)

For your tests I would suggest to manually make sure the downloaded binaries are in the output directory. Then the remaining question is if all bindings actually also work on Linux. I will also ask a college with a Linux machine to have a look at this (@hyazinthh). I hope this gets you one step further and we can work on making this library compatible with more platforms.

**Comment by ghost** [comment]
I actually did download the binaries as you said, and even put them in the folder. Aardvark did recognise them as native libraries, but it couldn't load them (was spitting out warnings) and the code still threw `DllNotFoundExceptions`. In the end, my solution was to extract the natives into the project, make them copy to the output on build, and then manually load them:
```csharp
Verbose("Loading OIDN lib manually");
		IntPtr oidnPtr = NativeLibrary.Load("oidn-natives/lib/libOpenImageDenoise.so"); //Have to manually load the lib or else it fails for some reason
		Verbose("OIDN Pointer is {Pointer}", oidnPtr);
		Verbose("Initializing Aardvark");
		Report.Verbosity = 999;
		Aardvark.Base.Aardvark.Init();
		//TODO: Serilog & Aardvark logs joined
		Verbose("Creating new OIDN Device");
		denoiseDevice = new Device();
		Verbose("OIDN Device created: {Device}", denoiseDevice);
```

This works now and I can see the denoising working fine!

**Comment by luithefirst** [comment]
Great to hear that to you could succeed and thanks for posting your workaround! If there is time we will have a more detailed look at the Linux support.

As I can see in your comment, you plan to join your logging and the Aardvark log output. I just wrote a quick start guide for the Aardvark logging functionality. You should be able to achieve this by implementing a custom `ILogTarget`: [aardvark.docs/Logging](https://github.com/aardvark-platform/aardvark.docs/wiki/Logging)

**Comment by hyazinthh** [comment]
There was an issue with the path in the `DllImport` attribute and the Linux binaries were missing as well. We have fixed those issues and released a new package (0.3.1). Works on my Ubuntu 22.04 system now.
